Prevent destructive action on production database

schneems opened the pull request:
This PR introduces a key/value type store to Active Record that can be used for storing internal values. It is an alternative implementation to #21237 cc @sgrif @matthewd.

It is possible to run your tests against your production database by accident right now. While infrequently, but as an anecdotal data point, Heroku receives a non-trivial number of requests for a database restore due to this happening. In these cases the loss can be large.

To prevent against running tests against production we can store the "environment" version that was used when migrating the database in a new internal table. Before executing tests we can see if the database is a listed in `protected_environments` and abort. There is a manual escape valve to force this check from happening with environment variable `DISABLE_DATABASE_ENVIRONMENT_CHECK=1`.

schneems committed:
Prevent destructive action on production database

This PR introduces a key/value type store to Active Record that can be used for storing internal values. It is an alternative implementation to #21237 cc @sgrif @matthewd.

It is possible to run your tests against your production database by accident right now. While infrequently, but as an anecdotal data point, Heroku receives a non-trivial number of requests for a database restore due to this happening. In these cases the loss can be large.

To prevent against running tests against production we can store the "environment" version that was used when migrating the database in a new internal table. Before executing tests we can see if the database is a listed in `protected_environments` and abort. There is a manual escape valve to force this check from happening with environment variable `DISABLE_DATABASE_ENVIRONMENT_CHECK=1`.

sgrif commented on activerecord/lib/active_record/internal_metadata.rb:
```diff
@@ -0,0 +1,56 @@
+
+      def store(hash)
+        hash.each do |key, value|
+          first_or_initialize(key: key.to_s).update_attributes!(value: value)
```
Why the `.to_s`? AR's type casting should handle this.

sgrif commented on activerecord/lib/active_record/internal_metadata.rb:
```diff
@@ -0,0 +1,56 @@
+      end
+
+      def value_for(key)
+        where(key: key.to_s).first.try(:value)
```
Could this be `where(key: key).pluck(:value).first`?

sgrif commented on activerecord/lib/active_record/internal_metadata.rb:
```diff
@@ -0,0 +1,56 @@
+      end
+
+      def table_exists?
+        ActiveSupport::Deprecation.silence { connection.table_exists?(table_name) }
```
What are we silencing? Should we make sure we're deprecation free here?

sgrif commented on activerecord/lib/active_record/internal_metadata.rb:
```diff
@@ -0,0 +1,56 @@
+          connection.remove_index table_name, name: index_name
+          connection.drop_table(table_name)
+        end
+      end
```
It doesn't look like this method is ever used.

sgrif commented on activerecord/lib/active_record/internal_metadata.rb:
```diff
@@ -0,0 +1,56 @@
+        ActiveSupport::Deprecation.silence { connection.table_exists?(table_name) }
+      end
+
+      # Creates a schema table with columns +environment+ and +version+
```
![b3qijlfccaejlwm](https://cloud.githubusercontent.com/assets/1529387/12184850/ec2c195a-b554-11e5-89c2-6b506efd69ec.jpg)

sgrif commented on activerecord/lib/active_record/migration.rb:
```diff
@@ -1200,12 +1221,26 @@ def record_version_state_after_migrating(version)
         migrated.delete(version)
         ActiveRecord::SchemaMigration.where(:version => version.to_s).delete_all
-      else
+       else
```
:scissors:

sgrif commented on activerecord/lib/active_record/internal_metadata.rb:
```diff
@@ -0,0 +1,56 @@
+        hash.each do |key, value|
+          first_or_initialize(key: key.to_s).update_attributes!(value: value)
+        end
+      end
```
Since this does one query per key, and it's only ever called with one key, what do you think about changing this to `[]=` and `value_for` to `[]`?

matthewd replied:
:+1:

schneems replied:
I have mixed feelings about this. I don't want it to feel like hash access because it is hitting the database. I want to avoid "Oh, look a hash, hashes are fast. We can put lots of stuff in it" reaction. I also don't want people to accidentally try to use multiple level keys etc. It doesn't look like a hash because it isn't one. I can change if you like `[]` better though.

matthewd replied:
That's fair. To me, absent a more comprehensive suite of hash-like methods, `[]`/`[]=` are just the "right" way of spelling "get/set value using a key", though.

And this is so far away from anything a user would ever touch, it seems fine to go with whatever makes our code look neatest.

sgrif replied:
Yeah, I think if this were public API that would be more of a concern. For
internal, I think [] and []= are fine.

That said, I don't feel particularly strongly on it, so I'll leave it up to
you since you did all the legwork on this.

On Thu, Jan 7, 2016 at 5:12 PM Matthew Draper <notifications@github.com>
wrote:

> In activerecord/lib/active_record/internal_metadata.rb
> <https://github.com/rails/rails/pull/22967#discussion_r49144499>:
>
> > +        "key"
> > +      end
> > +
> > +      def table_name
> > +        "#{table_name_prefix}#{ActiveRecord::Base.internal_metadata_table_name}#{table_name_suffix}"
> > +      end
> > +
> > +      def index_name
> > +        "#{table_name_prefix}unique_#{ActiveRecord::Base.internal_metadata_table_name}#{table_name_suffix}"
> > +      end
> > +
> > +      def store(hash)
> > +        hash.each do |key, value|
> > +          first_or_initialize(key: key.to_s).update_attributes!(value: value)
> > +        end
> > +      end
>
> That's fair. To me, absent a more comprehensive suite of hash-like
> methods, []/[]= are just the "right" way of spelling "get/set value using
> a key", though.
>
> And this is so far away from anything a user would ever touch, it seems
> fine to go with whatever makes our code look neatest.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rails/rails/pull/22967/files#r49144499>.
>

schneems committed:
[ci skip] Add comment to remove silenced code.

schneems committed:
Use `key` as primary key in schema.

schneems committed:
Add EnvironmentMismatchError

Raise an error when a destructive action is made on a database where the current environment is different from the environment stored in the database.

schneems committed:
Use hash like syntax for InternalMetadata

Discussion: https://github.com/rails/rails/pull/22967#discussion_r49137035

schneems commented:
I added the environment mismatch error in activerecord/lib/active_record/tasks/database_tasks.rb we're only checking it on "protected" rake tasks. I also added `[]` and `[]=` getter and setter methods

schneems committed:
Fix kwarg to not have circular dependency

schneems committed:
Fixing tests and re-locating error checking.

schneems commented:
All tests are :green_apple: :green_book: :green_heart:

schneems commented:
:rocket:

shunsukeaida commented on activerecord/lib/active_record/internal_metadata.rb:
```diff
@@ -0,0 +1,47 @@
+        ActiveSupport::Deprecation.silence { connection.table_exists?(table_name) }
+      end
+
+      # Creates a internal metadata table with columns +key+ and +value+
```
shouldn't it be `an internal`?

prathamesh-sonpatki replied:
Yes. Please send PR :)

shunsukeaida replied:
See, but #22992 beats me.

kipcole9 commented:
When migrating an app to master branch with this PR applied, a destructive task detects there is no environment metadata and advises running rake db:migrate to fix.

However the environment is only stored after a migration is run.  So if there are no pending migrations then the environment data is not added.  This is because record_version_state_after_migrating which does the actual update is called only from within execute_migration_in_transaction below.
````ruby
    def execute_migration_in_transaction(migration, direction)
      ddl_transaction(migration) do
        migration.migrate(direction)
        record_version_state_after_migrating(migration.version)
      end
    end
````
Therefore either the advise needs to change to an alternative strategy to update the environment metadata or rake db:migrate should alway update the environment metadata even when there are no pending migrations.  One approach might be to modify the migrate method:
````ruby
    def migrate
      if use_advisory_lock?
        with_advisory_lock { migrate_without_lock }
      else
        migrate_without_lock
      end
      ActiveRecord::InternalMetadata[:environment] = ActiveRecord::Migrator.current_environment
    end
````
and remove the update from record_version_state_after_migrating:
````ruby
    def record_version_state_after_migrating(version)
      if down?
        migrated.delete(version)
        ActiveRecord::SchemaMigration.where(:version => version.to_s).delete_all
      else
        migrated << version
        ActiveRecord::SchemaMigration.create!(version: version.to_s)
      end
    end
````
However it also means this method name is not quite on target and probably should be renamed record_version_after_migration again.

halilim commented:
Little late to the party and feeling like missing something obvious (sorry if this is the case), but shouldn't tests be allowed to run only on the test db? I can't think of a case I'd want to run tests on a non-test db and I also wouldn't like to accidentally destroy my development db by default for the sake of convenience.

So why not use a whitelist instead of a blacklist, e.g. `allow_environments`, or just remove it altogether?

schneems commented:
The core idea behind this feature is to protect environments that are not recoverable. Maybe running tests against your development database might suck, but it won't be company or career ending. To that end we're using a blacklist, you are declaring the things you want to protect. If you REALLY don't want tests run against your development database, you could add that in to the protected environment. 

You could always write your own whitelist

````
protected_environments = [Rails.env] unless UNPROTECTED_ENVIRONMENTS.include?(Rails.env)
```

halilim commented:
@schneems I thought `unprotected_environments` would be a better default by requiring less modifications for the common case. Thanks for answering.

JanStevens commented:
Also pretty late to the party but stumbling on this issue I find the implementation a bit overkill for what the ultimate goal is: Prevent destructive rake tasks in production.

The main disadvantage for this MR is:
- An extra internal table means that lots of gems will potentially brake (Apartment, SchemaPlus, to name a few. Let's not mention all the test runners).
- Using migrations to magically find out the env that we are currently using, storing it in an internal table and then checking if it matches our current running env seems quite overkill.
- There is no globally op out: I don't use heroku, lot's of people don't. I don't need this so why can't I op-out of the extra internal table?
- Not documented: One has to sift through the github MR + issues it caused to find out what to do. There is also no warning about a possible breaking change mentioned about this feature! Lot's of gems now get bug reports that they don't work with rails 5. Making overall transition to rails 5 harder.
- Recreating a test database with the combo `RAILS_ENV=test rake db:drop db:create db:schema:load` won't work. I need yet another rake task in the list, which makes no sense since I explicitly specify the env I'm running in.

For above reasons I propose the following alternative implementations:
-  A simple array `protected_envs` and then checking if current `RAILS_ENV` includes one of the protected_envs. As extra warning one could add a `are you sure?` confirmation.
- Just a confirmation asking the user if he/she is sure when in a protected_env. I think that this would help 90% of the use cases (if there are any outside heroku)
- Setting a required ENV variable before a destructive action can be taken place in a protected env
- Implement this in the CLI of heroku, so we don't pollute rails core with yet another basecamp/heroku feature.

If my alternative implementations won't do or I'm missing a very obvious part then please enlighten me.

schneems commented:
This feature is already in Rails. It's not coming out unless there's a really good reason. If you really want you can fork active record and remove it. There is already extensive conversation around alternative proposals and why they will or won't work. 

> An extra internal table means that lots of gems will potentially brake (Apartment, SchemaPlus, to name a few. Let's not mention all the test runners).

Open an issue with those gems or if there's a problem open an issue on Rails.

> Using migrations to magically find out the env that we are currently using, storing it in an internal table and then checking if it matches our current running env seems quite overkill.

No, we talked about it above, it's the only viable solution.

> There is no globally op out: I don't use heroku, lot's of people don't. I don't need this so why can't I op-out of the extra internal table?

That's not correct there's an env var you can set to allow you to disable the check. If you want to opt out of even having the table, you can fork active record.


> Not documented: One has to sift through the github MR + issues it caused to find out what to do. 

This is where you can help. What were you trying to do when you hit a problem? This feature __is__ documented, and claiming it isn't doesn't help us know what failed or why or when. The best docs come from people who use the product. Either open an issue with your failure case or submit a doc PR (an easy way to get on that leader board).

> There is also no warning about a possible breaking change mentioned about this feature! 

You're bumping major versions of Rails, which has lots of breaking changes. How exactly would you have like to be notified about this new feature as a breaking change?

> Lot's of gems now get bug reports that they don't work with rails 5. Making overall transition to rails 5 harder.

Again, specifics. We need specifics.

> Recreating a test database with the combo RAILS_ENV=test rake db:drop db:create db:schema:load won't work. I need yet another rake task in the list, which makes no sense since I explicitly specify the env I'm running in.


This is a really good piece of data. I don't use Rails like that, but you do. If you don't open an issue about these things we won't know about them. This really valuable piece of data is burried in your ranty comment on this pull request. I didn't like your tone, so honestly I didn't even read it until today.

If you want to make progress here, please open an issue with that as the text. You can even include example output:

```
$ RAILS_ENV=test rake db:drop db:create db:schema:load
DEPRECATION WARNING: `config.serve_static_files` is deprecated and will be removed in Rails 5.1.
Please use `config.public_file_server.enabled = true` instead.
 (called from block in <top (required)> at /Users/richardschneeman/Documents/projects/tmp/loggertest/config/environments/test.rb:16)
DEPRECATION WARNING: `config.static_cache_control` is deprecated and will be removed in Rails 5.1.
Please use
`config.public_file_server.headers = { 'Cache-Control' => 'public, max-age=3600' }`
instead.
 (called from block in <top (required)> at /Users/richardschneeman/Documents/projects/tmp/loggertest/config/environments/test.rb:17)
DEPRECATION WARNING: ActiveRecord::Base.raise_in_transactional_callbacks= is deprecated, has no effect and will be removed without replacement. (called from <top (required)> at /Users/richardschneeman/Documents/projects/tmp/loggertest/config/environment.rb:5)
rake aborted!
ActiveRecord::NoEnvironmentInSchemaError:

Environment data not found in the schema. To resolve this issue, run:
```

> A simple array protected_envs and then checking if current RAILS_ENV includes one of the protected_envs. As extra warning one could add a are you sure? confirmation.

Won't work. Also confirmations break automated work flows and scripts.

> Just a confirmation asking the user if he/she is sure when in a protected_env. I think that this would help 90% of the use cases (if there are any outside heroku)

Cconfirmations break automated work flows and scripts. We essentially have this by allowing you to manage your experience via environment variables. We even list how to proceed if you really want to in the error message

```
Environment data not found in the schema. To resolve this issue, run:

	bin/rails db:environment:set RAILS_ENV=test
```

> Setting a required ENV variable before a destructive action can be taken place in a protected env

Won't work.

> Implement this in the CLI of heroku, so we don't pollute rails core with yet another basecamp/heroku feature.

schneems commented:
I appreciate your desire to be involved with the process. I'm frustrated by the amount of time i'm having to re-explain myself over the mailing list and github comments over this issue. I just got back from vacation and would much rather spend time fixing bugs, or writing new features, or cutting releases of sprockets rather than waste so much time on github comments.

I've pointed out some productive avenues for your experiences. I really appreciate you bringing the failure case to our attention. You can open a new issue and mention be @schneems and I can take a look at it later.
